File: p114.java
/* 
 * Solution to Project Euler problem 114
 * By Nayuki Minase
 * 
 * http://nayuki.eigenstate.org/page/project-euler-solutions
 * https://github.com/nayuki/Project-Euler-solutions
 */


public final class p114 {
	
	public static void main(String[] args) {
		long[] ways = new long[51];  // Memoization
		ways[0] = 1;
		ways[1] = 1;
		ways[2] = 1;
		for (int i = 3; i <= 50; i++) {
			ways[i] += ways[i - 1];
			for (int j = 3; j < i; j++)
				ways[i] += ways[i - j - 1];  // Dynamic programming
			ways[i] += ways[0];
		}
		System.out.println(ways[50]);
	}
	
}

File: haskell/p114.hs
{- 
 - Solution to Project Euler problem 114
 - Copyright (c) Project Nayuki. All rights reserved.
 - 
 - https://www.nayuki.io/page/project-euler-solutions
 - https://github.com/nayuki/Project-Euler-solutions
 -}


{- 
 - How many ways can a row n units long be filled? Denote this quantity as ways[n].
 - Compute n = 0, 1, 2 manually as base cases.
 - 
 - Now assume n >= 3. Look at the leftmost item and sum up the possibilities.
 - * If the item is a grey square, then the rest of the row is allowed
 -   to be anything of length n-1. Add ways[n-1].
 - * If the item is a red block with length k where k >= 3, then:
 -   * If k = n, then the whole row is filled by this red block. Add 1.
 -   * Otherwise k < n, this red block is followed by a grey square, then followed
 -     by anything of length n-k-1. So add ways[n-4] + ways[n-5] + ... + ways[0].
 -}
main = putStrLn (show ans)
ans = ways 50

ways 0 = 1
ways 1 = 1
ways 2 = 1
ways n = (waysMemo !! (n-1)) + 1 + (sum (take (n-3) waysMemo))
waysMemo = map ways [0..]
